android/kotlin/v2/AmpliApp/app/src/main/java/com/amplitude/ampli/IngestionMetadataErichmentPlugin.kt

```diff
@@ -0,0 +1,49 @@
+package com.amplitude.ampli
+
+import com.amplitude.core.events.BaseEvent
+import com.amplitude.core.events.IngestionMetadata
+import com.amplitude.core.platform.Plugin
+import java.util.regex.Pattern
+
+var minIngestionMetadataSupportedVersion = "1.2.0";
+class IngestionMetadataErichmentPlugin : Plugin{
+    override val type: Plugin.Type = Plugin.Type.Enrichment
+    override lateinit var amplitude: com.amplitude.core.Amplitude
+
+    private fun compare(v1: String, v2: String) : Int {
+        val s1 = normalisedVersion(v1)
+        val s2 = normalisedVersion(v2)
+        val cmp = s1.compareTo(s2)
+        return cmp
+    }
+
+    fun normalisedVersion(version: String?): String {
+        return normalisedVersion(version, ".", 4)
+    }
+
+    fun normalisedVersion(version: String?, sep: String?, maxWidth: Int): String {
+        val split: Array<String> = Pattern.compile(sep, Pattern.LITERAL).split(version)
+        val sb = StringBuilder()
+        for (s in split) {
+            sb.append(String.format("%" + maxWidth + 's', s))
+        }
+        return sb.toString()
+    }
+
+    override fun execute(event: BaseEvent): BaseEvent? {
+
+        var libraryArr = event.library?.split("/")
+        var libraryVersion = libraryArr?.get(1)
+        println(libraryVersion)
+        println("####################")
+        if (!libraryVersion.isNullOrEmpty() && compare(libraryVersion, minIngestionMetadataSupportedVersion) >= 0) {
+            println("should set the ingestionMetadata")
+            println(libraryVersion)
+            event.ingestionMetadata
```

Comment:
Will this give a compilation error in older SDKs? We need reflection or another way to dynamically access this property.

Reply:
Similar discussion with @liuyang1520 on this earlier. Maybe try on the earlier versions and verify it is not throwing compilation error.

Reply:
Doing the investigation about the dynamic access.

Reply:
We can instead set `event.extra.amplitude = IngestionMetadata()` and then read it in Amplitude Destination Plugin.
https://github.com/amplitude/Amplitude-Kotlin/blob/main/core/src/main/java/com/amplitude/core/events/EventOptions.kt#L42

Reply:
Or we could do the JSON in core:

Plugin
`event.extra["amplitude"] = { ingestionMetadata: IngestionMetadata() }`

Core:
`val ingestiontionJson = JsonObject(event.extra["amplitude"]["ingestionMetadata"])`

Reply:
This approach may be better since it doesn't require reflection (slow). WDYT? @yuhao900914 @qingzhuozhen @liuyang1520